Adding slack integration for posting to channel on applies. Fixes #179

nicholas-wu-hs opened the pull request:
Tackling #179, first iteration using `slack-token` and `slack-channel` in the config.
If specified, a Slack Client will be created and attached to `ApplyExecutor`. It's used to post messages to `slack-channel` on applies (both success and fail).

Using https://github.com/nlopes/slack for slack client.

Will probably need a lot of refactoring/restructuring, then eventually tests and docs?
Also, was I supposed to add the dependencies into the repo?

nicholas-wu-hs commented on server/server.go:
```diff
@@ -61,6 +64,14 @@ func NewServer(config Config) (*Server, error) {
 	githubStatus := &events.GithubStatus{Client: githubClient}
+	// nil slackClient unless token and channel was specified
+	var slackClient slack.Client
+	if config.SlackToken != "" && config.SlackChannel != "" {
```
Only attempt to create a slack client if both `token` and `channel` are supplied.
Should probably add a warning if they only supply one of them IF we decide to keep this schema, but it will probably be changed after discussion in #179

lkysow replied:
In general it's dangerous to use `nil` for something like this because it's not obvious that my slack client will be nil just because somewhere up the chain depending on the config. On the other hand, you can't initialize a slack client if you don't have the config :D.

I would create a NoopSlackClient that implements the interface but doesn't do anything and use it here if there is no slack config.

lkysow commented on server/events/apply_executor.go:
```diff
@@ -106,3 +114,19 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
 	return ProjectResult{ApplySuccess: output}
 }
+
+func createSlackMessage(ctx *CommandContext, success bool) string {
```
this functionality doesn't belong in the apply executor since it's all about applying, not creating slack messages. Put this in the slack client.

lkysow replied:
don't use package functions like this, instead add it to the struct.
`func (a *ApplyExecutor) createSlackMessage`

If you don't do this then anyone can call this method from anywhere in this package which is weird, you really only want it called from the object

lkysow commented on server/events/apply_executor.go:
```diff
@@ -106,3 +114,19 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
+		status = ":x:"
+	}
+
+	return fmt.Sprintf("%s *%s* %s in <%s|%s>.",
```
can you add a screenshot so we can see what this looks like?

nicholas-wu-hs replied:
Here it is for using the emojis!
![screen shot 2017-11-02 at 4 25 28 pm](https://user-images.githubusercontent.com/31903040/32355209-a89fd86e-bfea-11e7-8752-62d0ee1ac433.png)

nicholas-wu-hs replied:
Also had them as text at one point:
![screen shot 2017-11-02 at 4 25 45 pm](https://user-images.githubusercontent.com/31903040/32355556-989f37aa-bfec-11e7-83a5-9fbfed05f1b0.png)

nicholas-wu-hs replied:
And as attachments but they're chunkier:
![screen shot 2017-11-02 at 4 42 09 pm](https://user-images.githubusercontent.com/31903040/32355665-515896ce-bfed-11e7-8081-aadf6fb8a68e.png)

lkysow commented on server/events/slack/slack_client.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	// https://api.slack.com/faq
+	// 'How do I find a channel's ID if I only have its #name?'
```
I don't think you need this. It says you can just use the name: https://api.slack.com/methods/chat.postMessage#channels

nicholas-wu-hs replied:
With `chat.postMessage` you can use a channel name, but if the channel doesn't exist, it doesn't send.

I was thinking of making sure the `channel` from config exists on the creation of the slack client and failing early if it doesn't, rather than failing at `postMessage` or failing silently.

So to do that, I looked at https://api.slack.com/methods/channels.info which requires an ID. Alternatively, I can do a `postMessage` and see if it fails but that would post a message?

lkysow replied:
I like the idea of failing when Atlantis server starts if that channel doesn't exist. Failing early is usually better so good call.

lkysow commented on server/events/slack/slack_client.go:
```diff
@@ -0,0 +1,52 @@
+		}
+	}
+
+	return nil, errors.New("channel_not_found")
```
errors should be human readable so you could have just had "channel doesn't exist"

nickwu241 commented:
Thanks for the awesome review and feedback!
- Restructured to be more generic and handle Webhook config schema discussed at #179
- Fixed errors to be human readable
- Refactored slack related items to the `server/events/slack`
- Reverted 3be7f8c (adding vendor deps) for now so it's easier to review

Still have a few more things to do:
- [ ] Proper error handling and messages
- [ ] Check if slack channel exists on server startup and error early if it doesn't
- [ ] ?Default `workspace-regex` to `".*"` if not provided
- [ ] ?Pass command success/failure status down to `HookExecutor`s
- [ ] Add tests

nickwu241 commented on server/server.go:
```diff
@@ -44,18 +45,73 @@ type Server struct {
+	for _, webhook := range config.Webhooks {
+		hookExecutor, err := hookExecutorBuilder.build(webhook)
+		if err != nil {
+			// TODO: check if it was because of uninitialized slack client
```
If `slack-token` wasn't provided in the config, `hookExecutorBuilder` would have a `nil` `slack.Client`. This causes an error when building slack webhooks, then Atlantis can error saying the user needs to provide a `slack-token`.

Using a `NoopClient` would work, but if we can error early and tell the user what's wrong, I feel like that's the way to go! But this doesn't seem too clean, any suggestions?

lkysow replied:
In my suggestions I have a `WebhooksManager` that has a `[]WebhookSender`. When it gets constructed with no webhooks then it just has an empty array and so calling `WebhooksManager.Send()` just does nothing. That's how I handled it.

When it's running through the webhooks in its constructor, it returns an `error` if a slack webhook was specified but there was no `slack-token`: https://github.com/nicholas-wu-hs/atlantis/pull/1/files?diff=unified#diff-0250d5b43682815402ad30ce1b068cdcR55

nickwu241 commented on server/events/apply_executor.go:
```diff
@@ -92,6 +94,19 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
+		matched, err := regexp.MatchString(workspaceRegex, ctx.Command.Environment)
+		if err != nil {
+			// log the regexp error but let's continue with the apply
+			ctx.Log.Debug(err.Error())
```
Maybe we can check during the config processing stage to verify a valid regex?
Does a debug message make sense here? Or should it be an error :O

lkysow replied:
> Maybe we can verify the regex during the config processing stage?

exactly `regexp.Compile`

lkysow commented:
I'll add specific comments to your code but in general here's the changes I think need to be made and why:
- lets move everything under an `events/webhooks` package. I don't think slack needs its own package at the moment and then all our code can be under that package cleanly including `hook_executor`
- I want the code in `ApplyExecutor` to be as clean as possible and not need to know anything about webhooks. The ideal call would be just be
```go
	a.WebhooksManager.Send(webhookResult)
```
- constructing all the webhooks clients and verifying the regex, etc. should happen inside the `WebhooksManager` constructor. So in `server/server.go` it should just look like
```go
	webhookManager, err := webhooks.NewWebhooksManager(webhooksConfig, config.SlackToken)
	if err != nil {
		return nil, errors.Wrap(err, "initializing webhooks")
	}
```
- the interfaces I'd like to see are
  - a `WebhooksSender` that handles sending out multiple webhooks. This is what the `ApplyExecutor` has a reference to.
```go
type WebhooksSender interface {
	Send(log *logging.SimpleLogger, result ApplyResult)
}
```
  - a `WebhookSender` that handles sending out a *single* webhook. This can either be implemented by the `SlackSender` or a `GenericSender`. The `WebhooksSender` will have a reference to multiple `WebhookSender`'s. These will decide based on the environment whether to send out their webhooks or not.
```go
type WebhookSender interface {
	Send(ApplyResult) error
}
```

lkysow commented on server/server.go:
```diff
@@ -44,18 +45,73 @@ type Server struct {
+	Channel string `mapstructure:"channel"`
+}
+
+type HookExecutorBuilder struct {
```
I would all of this code out of the `server` package and into `events/webhooks`. Ask yourself if building a HookExecutor should be the responsibility of the server class

nicholas-wu-hs replied:
Good point, I thought about this too. 
Previously, I tried to pass the config directly to the `server/events/webhooks` package (thus needing to import `server`) and ran into circular import problems, googled a bit for more info but no clear solution.

In general, does a `.go` file import packages from parent directories?

lkysow replied:
no it doesn't because of those circular import errors you ran into. You can either create a separate package (like `models`) or you only reference from parent to child, never from child to parent

nicholas-wu-hs replied:
Got it, thank you 👍

lkysow commented on server/server.go:
```diff
@@ -44,18 +45,73 @@ type Server struct {
+	Slack slack.Client
+}
+
+func (builder *HookExecutorBuilder) build(webhook Webhook) (events.HookExecutor, error) {
```
pointer receivers should just be the first letter of the struct, in this case `h *HookExecutorBuilder`.

nicholas-wu-hs replied:
Got it now! Thanks, will do this for all future code!

lkysow commented on server/server.go:
```diff
@@ -86,6 +142,7 @@ func NewServer(config Config) (*Server, error) {
 	}
 	applyExecutor := &events.ApplyExecutor{
 		Github:            githubClient,
+		WSRegexToHook:     applyExecutorRegexToHook,
```
here's where I would have liked to see something like a `WebhooksManager` so the ApplyExecutor only needs to call a single line of code to send webhooks

lkysow commented on server/server.go:
```diff
@@ -44,18 +45,73 @@ type Server struct {
+	Webhooks            []Webhook `mapstructure:"webhooks"`
+}
+
+type Webhook struct {
```
I'd call this `WebhookConfig` since it's no an actual webhook, just config for it.

In my suggestions I don't pass this in directly to the `WebhooksManager`. This is because I want the config that users need in their config files to be decoupled from the WebhooksManager. This will prevent us from accidentally changing the name of a config option in the `WebhooksManager` code and invalidating everyone's config files :) We'll have to explicitly do the mapping: https://github.com/nicholas-wu-hs/atlantis/pull/1/files?diff=unified#diff-91bbeda7eb98a7adc57b9e47e2cf5c2bR72

lkysow commented on server/events/apply_executor.go:
```diff
@@ -92,6 +94,19 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
 	output, err := a.Terraform.RunCommandWithVersion(ctx.Log, absolutePath, tfApplyCmd, terraformVersion, env)
+
+	// execute apply webhooks
+	for workspaceRegex, hookExecutor := range a.WSRegexToHook {
```
ideally this is just one line call to `WebhooksManager`

nicholas-wu-hs replied:
I see! I agree this is the way to go to keep the code in `ApplyExecutor` as clean as possible and more reusable, it makes it easy to extend this functionality to other `Executor`s in the future.

lkysow commented on server/events/apply_executor.go:
```diff
@@ -92,6 +94,19 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
+
+	// execute apply webhooks
+	for workspaceRegex, hookExecutor := range a.WSRegexToHook {
+		matched, err := regexp.MatchString(workspaceRegex, ctx.Command.Environment)
```
if you had constructed the regex during the construction of the slack webhook sender then here you could have a reference to `*regexp.Regexp` and call `.MatchString()` which won't return an error.

nicholas-wu-hs replied:
Awesome! Thanks, still new to golang :P

lkysow commented on server/events/slack/slack_client.go:
```diff
@@ -0,0 +1,61 @@
+	client *slack.Client
+}
+
+type SlackMessageTemplate struct {
```
good call to have this as a struct rather than multiple params

lkysow commented on server/events/slack/slack_client.go:
```diff
@@ -0,0 +1,61 @@
+	}, nil
+}
+
+func (s *ConcreteClient) PostMessage(channel string, template SlackMessageTemplate) (string, error) {
```
here you should have `(c *ConcreteClient)`

lkysow commented on server/events/slack/slack_client.go:
```diff
@@ -0,0 +1,61 @@
+	return timestamp, err
+}
+
+func (s *ConcreteClient) createMessage(template SlackMessageTemplate) string {
```
I want to do the attachment style message that you showed earlier. I think that looks best.

nicholas-wu-hs replied:
Agreed that it looks nicer, one downside is that it takes a bit more room (one attachment is at min ~3 lines), I'll play around a bit more with the message style.

lkysow replied:
yeah it takes up more room but I think it's worth it because it allows the information to be shown in a more consumable manner

nicholas-wu-hs commented:
Still working on the tests, looking into using `pegomock`, I'll let you know when this is ready for another review :)

So this is another shot at using slack attachments, what do you think? 
- is `Environment` important enough to keep as its own field
- do we need a `Status` field? it's already encoded with the colors so I think no
![screen shot 2017-11-10 at 9 58 05 am](https://user-images.githubusercontent.com/31903040/32671855-d25945c8-c5fd-11e7-83e2-3e6e91deef87.png)

- if we want the attachment to be smaller, we can do something like `Applied <ENV> in <REPO> by <USER>.` or `<USER> applied <ENV> in <REPO>.`
![screen shot 2017-11-10 at 10 07 01 am](https://user-images.githubusercontent.com/31903040/32672180-ed2e5612-c5fe-11e7-9a92-8f5f6c39fe6d.png)

lkysow commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+	configs := validConfigs()
+	configs[0].Event = unsupportedEvent
+	_, err := webhooks.NewWebhooksManager(configs, validToken)
+	Assert(t, strings.Contains(err.Error(), "event"), "expected err")
```
I don't think it's enough just to look for the word `event` here. I'd like to see you matching on the whole error
`Equals(t, "event: badevent not supported. Only event: apply is supported right now", err.Error())`. This will ensure we're including the event name that wasn't supported in the error which is important for users since otherwise they won't know what event was wrong

lkysow commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+	configs := validConfigs()
+	configs[0].Kind = unsupportedKind
+	_, err := webhooks.NewWebhooksManager(configs, validToken)
+	Assert(t, strings.Contains(err.Error(), "kind"), "expected err")
```
use `Equals` and check the whole error string

lkysow commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+	Channel:        validChannel,
+}
+
+func validConfigs() []webhooks.Config {
```
what is this doing? Why are you appending an empty slice to `validConfig`? Why not just have
```
var validConfig = []webhooks.Config{
     {Event: ...}
}
```

nicholas-wu-hs replied:
What I wanted to do was get a fresh copy of the valid configs for each test since I modify it in some tests.
So I think after having `validConfig` as a variable, I can do `config := validConfig` to copy it.

nicholas-wu-hs replied:
I left it as a function because I want new copies of the valid configs in each test, but I got rid of the appending.

lkysow commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+func TestNewWebhooksManager_NoSlackToken(t *testing.T) {
+	emptyToken := ""
+	_, err := webhooks.NewWebhooksManager(validConfigs(), emptyToken)
+	Assert(t, strings.Contains(err.Error(), "slack-token must be set"), "expected err")
```
we should also test that if there are no configs then you don't need a `slack-token`

lkysow commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+	// todo: failing because not a valid token and channel
+	m, err := webhooks.NewWebhooksManager(validConfigs(), validToken)
+	Ok(t, err)
+	Assert(t, m != nil, "mangager shouldn't be nil")
```
mangager => manager

nicholas-wu-hs replied:
thanks for the catch!

lkysow commented on server/server.go:
```diff
@@ -44,18 +45,37 @@ type Server struct {
 func NewServer(config Config) (*Server, error) {
+	var webhooksConfig []webhooks.Config
+	for _, c := range config.Webhooks {
+		webhooksConfig = append(webhooksConfig, webhooks.Config{Channel: c.Channel, Event: c.Event, Kind: c.Kind, WorkspaceRegex: c.WorkspaceRegex})
```
let's use a multiline config for the struct here since this line is getting long

nicholas-wu-hs commented on server/events/apply_executor.go:
```diff
@@ -92,6 +94,16 @@ func (a *ApplyExecutor) apply(ctx *CommandContext, repoDir string, plan models.P
 	tfApplyCmd := append(append(append([]string{"apply", "-no-color"}, applyExtraArgs...), ctx.Command.Flags...), plan.LocalPath)
 	output, err := a.Terraform.RunCommandWithVersion(ctx.Log, absolutePath, tfApplyCmd, terraformVersion, env)
+
+	// Send webhooks.
```
Thanks, I agree with the view on what comments should be and to strive to write self-documenting code -- will remove that line.

nicholas-wu-hs commented on server/events/webhooks/webhooks_test.go:
```diff
@@ -0,0 +1,84 @@
+	configs := validConfigs()
+	configs[0].WorkspaceRegex = invalidRegex
+	_, err := webhooks.NewWebhooksManager(configs, validToken)
+	Assert(t, strings.Contains(err.Error(), "error parsing regexp"), "expected err")
```
That makes sense, now I know and will `nil` check errors in the tests now & for the future, ty!

nicholas-wu-hs commented on server/events/webhooks/mocks/mock_slack.go:
```diff
@@ -0,0 +1,151 @@
+	return ret0, ret1
+}
+
+func (mock *MockSlackClient) PostMessage(channel string, result ApplyResult) error {
```
the generated code doesn't compile because it doesn't know what `ApplyResult` is.
I tried adding in the package but then got errors about `pegomock.GetGenericMockFrom(mock).Invoke` return type is incompatible.
Still figuring this out!

nicholas-wu-hs commented:
Still figuring out the mocks, but so far I've
- changed `color` => `colour`, move red, green to constants
- fixed `mangager` => `manager`
- removed redundant comment
- added log message for most tests (still WIP)
- added `nil` checks before `err.Error()` in tests
- tested for whole strings for `event` and `kind`
- refactored `slack.go` so it can be mocked

I'll ask for a review after I'm done but wouldn't mind another review :)

lkysow commented on server/events/webhooks/slack.go:
```diff
@@ -0,0 +1,121 @@
+	failureColour = "danger"
+)
+
+//go:generate pegomock generate --use-experimental-model-gen --package mocks -o mocks/mock_slack.go slack.go
```
change `slack.go` to `SlackClient`. Follow method 2. here: https://github.com/petergtz/pegomock#generating-mocks

nicholas-wu-hs replied:
Wow, thank you ahahaha, that fixed the can't find `ApplyResult`
Now I'm looking into
```
mocks/mock_slack.go:50:9: no new variables on left side of :=
mocks/mock_slack.go:50:9: cannot use pegomock.GetGenericMockFrom(mock).Invoke("PostMessage", params, []reflect.Type literal) (type pegomock.ReturnValues) as type webhooks.ApplyResult in assignment
mocks/mock_slack.go:52:8: invalid argument result (type webhooks.ApplyResult) for len
mocks/mock_slack.go:53:12: invalid operation: result[0] (type webhooks.ApplyResult does not support indexing)
mocks/mock_slack.go:54:17: invalid operation: result[0] (type webhooks.ApplyResult does not support indexing)
```

nicholas-wu-hs replied:
as you mentioned, this was due to variable name clashing so I changed `result` to `applyResult` on the interface, thanks for catching this :)

lkysow commented on server/events/webhooks/webhooks.go:
```diff
@@ -0,0 +1,76 @@
+const SlackKind = "slack"
+const ApplyEvent = "apply"
+
+//go:generate pegomock generate --use-experimental-model-gen --package mocks -o mocks/mock_webhooks.go webhooks.go
```
change `webhooks.go` to `WebhooksSender`

nicholas-wu-hs commented:
- fixed the mocks per your feedback
- modified tests to use mock slack client
- added tests to `slack_test.go`, all tests are passing!

lkysow commented:
Your commit isn't passing `gometalint`.
```
# gotype and gotypex are disabled because they don't pass on CI and https://github.com/alecthomas/gometalinter/issues/206
# gocyclo is temporarily disabled because we don't pass it right now
# golint is temporarily disabled because we need to add comments everywhere first
gometalinter --disable gotype --disable gotypex --disable=gocyclo --disable golint --enable=megacheck --enable=unparam --deadline=120s --vendor -t --line-length=120 $(find . -type f -name '*.go' ! -path "./vendor/*" ! -path "./server/static/bindata_assetfs.go" ! -path "**/mocks/*" | xargs -I '{}' dirname '{}' | sort -u)
server/events/webhooks/slack_test.go:29:11:warning: error return value not checked (hook.Send(result)) (errcheck)
Makefile:51: recipe for target 'gometalint' failed
```

nicholas-wu-hs commented:
I see! I think I know the problem from the error message, will add in the fixes.
Should I/do I need to add in the vendor dependencies (i.e. the slack library) as part of this pull request?

nicholas-wu-hs commented on server/events/webhooks/slack_test.go:
```diff
@@ -0,0 +1,50 @@
+		Environment: "production",
+	}
+	err = hook.Send(result)
+	Ok(t, err)
```
Yeah, you're correct, the test passes commenting out the regex match.
`client.VerifyWasCalled(Never()).PostMessage(...)` works and the test fails if i comment out the regex match! ty ty

nicholas-wu-hs commented:
Finished writing the tests

lkysow commented:
Merged in #199!
